File: src/ontocopilot/onto/engagement.py
"""Frozen FDE engagement workflow declaration.

The executable handlers live in :mod:`.engagement_runtime`; keeping the declaration
separate preserves the stable control-plane contract.  Material *contents* are node
inputs and can never add tools or alter the topology.
"""

from __future__ import annotations

from ..kernel.agents import AgentLibrary, default_agents
from ..kernel.dag import (
    Dag,
    Difficulty,
    GateSpec,
    NodeBudget,
    NodeMode,
    NodeSpec,
    ScopeSpec,
)

FDE_ENGAGEMENT_AGENTS: tuple[str, ...] = (
    "fde_interviewer",
    "process_modeler",
    "erp_mapper",
    "rule_engineer",
    "data_steward",
    "delivery_reviewer",
)


def _agent_node(
    node_id: str,
    agent_name: str,
    *,
    deps: tuple[str, ...] = (),
    evidence_top_k: int = 32,
    gate: GateSpec | None = None,
) -> NodeSpec:
    """Translate an :class:`AgentSpec` into the existing DAG node contract."""
    agent = default_agents().get(agent_name)
    return NodeSpec(
        id=node_id,
        mode=agent.mode,
        handler=f"agent.{agent_name}",
        deps=deps,
        scope=ScopeSpec(evidence_top_k=evidence_top_k),
        budget=agent.budget,
        critics=agent.critics,
        critic_rounds=agent.critic_rounds,
        difficulty=agent.difficulty,
        gate=gate,
        params={
            "agent": agent_name,
            "tool_scope": agent.tool_scope,
            "output_schema": agent.output_schema,
        },
    )


def build_fde_engagement_dag(
    agents: AgentLibrary | None = None,
) -> Dag:
    """Build and freeze the front-line discovery-to-delivery workflow.

    The optional ``agents`` argument is a fail-fast compatibility check for deployments
    that extend the default agent library.  Nodes still carry only serializable names and
    schemas, so the declaration can be journalled or rendered without live agent objects.

    Topology::

        INTAKE -> PROCESS -> ERP_MAP -----\
                          -> RULES --------> GAP -> INTERVIEW (HITL)
                          -> DATA_OBJECTS -/             |
                                             CANONICALIZE -> REVIEW -> EXPORT

    ``GAP`` is the synchronization barrier.  ``INTERVIEW`` is always present even when the
    current backlog is empty; its deterministic handler may immediately accept an empty
    answer.  This keeps uploaded content from changing the plan after it is frozen.
    """
    library = agents or default_agents()
    for name in FDE_ENGAGEMENT_AGENTS:
        library.get(name)

    dag = Dag("fde_engagement_v1", freeze_before="INTAKE")
    dag.extend([
        _agent_node("INTAKE", "fde_interviewer", evidence_top_k=40),
        _agent_node("PROCESS", "process_modeler", deps=("INTAKE",), evidence_top_k=48),
        _agent_node("ERP_MAP", "erp_mapper", deps=("PROCESS",), evidence_top_k=40),
        _agent_node("RULES", "rule_engineer", deps=("PROCESS",), evidence_top_k=40),
        _agent_node(
            "DATA_OBJECTS", "data_steward", deps=("PROCESS",), evidence_top_k=40,
        ),
        NodeSpec(
            id="GAP",
            mode=NodeMode.DETERMINISTIC,
            handler="engagement.collect_gaps",
            deps=("PROCESS", "ERP_MAP", "RULES", "DATA_OBJECTS"),
            scope=ScopeSpec(evidence_top_k=0, recall_long_term=False),
            budget=NodeBudget(tokens=0, iterations=1, wallclock_s=60, tool_calls=0),
            params={
                "output_contract": "QuestionBacklog",
                "rank_by": (
                    "downstream_blocking", "blast_radius", "irreversibility", "evidence_gap",
                ),
            },
            retries=0,
        ),
        NodeSpec(
            id="INTERVIEW",
            mode=NodeMode.HITL,
            handler="engagement.interview",
            deps=("GAP",),
            scope=ScopeSpec(evidence_top_k=0, recall_long_term=False),
            budget=NodeBudget(tokens=0, iterations=1, wallclock_s=604_800, tool_calls=0),
            difficulty=Difficulty.LOW,
            params={
                "input_contract": "QuestionBacklog",
                "output_contract": "DecisionLedger",
                "batching": "progressive",
            },
            retries=0,
        ),
        NodeSpec(
            id="CANONICALIZE",
            mode=NodeMode.DETERMINISTIC,
            handler="engagement.canonicalize",
            deps=("INTERVIEW",),
            scope=ScopeSpec(evidence_top_k=0),
            budget=NodeBudget(tokens=0, iterations=1, wallclock_s=120, tool_calls=0),
            params={"output_contract": "OntologyPackage.v1"},
            retries=0,
        ),
        _agent_node(
            "REVIEW", "delivery_reviewer", deps=("CANONICALIZE",), evidence_top_k=24,
            gate=GateSpec(
                kind="auto",
                require=(
                    "verdict == 'PASS'", "blocker_count == 0",
                    "all_passed == true", "high_findings == 0",
                ),
            ),
        ),
        NodeSpec(
            id="EXPORT",
            mode=NodeMode.DETERMINISTIC,
            handler="engagement.export",
            # Export must receive both the reviewed package and the review verdict.
            # A transitive dependency is not included in WorkingSet.select(); declaring
            # both inputs prevents the release handler from silently re-validating `{}`.
            deps=("CANONICALIZE", "REVIEW"),
            scope=ScopeSpec(evidence_top_k=0, recall_long_term=False),
            budget=NodeBudget(tokens=0, iterations=1, wallclock_s=180, tool_calls=0),
            gate=GateSpec(
                kind="auto",
                require=(
                    "review_passed == true", "schema_valid == true",
                    "downloadable == true",
                ),
            ),
            params={
                "formats": ("json", "xlsx", "md", "mermaid"),
                "input_contract": "OntologyPackage.v1",
            },
            retries=0,
        ),
    ])
    return dag.freeze()


__all__ = ["FDE_ENGAGEMENT_AGENTS", "build_fde_engagement_dag"]

File: ts/src/onto/engagement.ts
/**
 * Compatibility facade for the catalog-backed, frozen FDE engagement workflow.
 *
 * The executable handlers live in engagement_runtime.ts.  The control-plane topology
 * lives in ts/catalog/workflows/fde-engagement.yaml and is loaded once at startup; user
 * material can never select another file or mutate the frozen graph.
 */

import { type AgentLibrary, defaultAgents } from "../kernel/agents.js";
import { NodeMode } from "../kernel/dag.js";
import {
  buildWorkflowDag,
  loadWorkflowDefinition,
  type WorkflowDefinition,
} from "../workflows/loader.js";

const FDE_DEFINITION = loadWorkflowDefinition("fde-engagement.yaml");

function assertFdeSafetyContract(definition: WorkflowDefinition): void {
  if (definition.name !== "fde_engagement_v3") {
    throw new Error(`FDE workflow name 不可变更: ${definition.name}`);
  }
  if (definition.freezeBefore !== "INTAKE" || definition.nodes[0]?.id !== "INTAKE") {
    throw new Error("FDE workflow 必须在首个 INTAKE 节点前冻结");
  }
  const interview = definition.nodes.find((node) => node.id === "INTERVIEW");
  if (
    interview?.kind !== "handler" ||
    interview.mode !== NodeMode.HITL ||
    interview.handler !== "engagement.interview" ||
    interview.gate === null ||
    !interview.gate.require.includes("contract == 'DecisionLedger'") ||
    !interview.gate.require.includes("resolved == true")
  ) {
    throw new Error("FDE INTERVIEW 必须保留 DecisionLedger 人工门");
  }
  const decisionProposal = definition.nodes.find((node) => node.id === "DECISION_PROPOSAL");
  if (
    decisionProposal?.kind !== "agent" ||
    decisionProposal.agent !== "decision_integrator"
  ) {
    throw new Error("FDE DECISION_PROPOSAL 必须由 decision_integrator 生成只读变更建议");
  }
  const decisionApply = definition.nodes.find((node) => node.id === "DECISION_APPLY");
  if (
    decisionApply?.kind !== "handler" ||
    decisionApply.mode !== NodeMode.DETERMINISTIC ||
    decisionApply.handler !== "engagement.validate_decision_application" ||
    decisionApply.gate === null ||
    !decisionApply.gate.require.includes("untargeted_write_count == 0") ||
    !decisionApply.gate.require.includes("mutation_count == 0") ||
    !decisionApply.gate.require.includes("claimed_applied_count == 0")
  ) {
    throw new Error("FDE DECISION_APPLY 必须确定性、无写权限并对无目标/伪应用 fail closed");
  }
  for (const [id, agent] of [
    ["REQUIREMENTS", "requirements_engineer"],
    ["ARCHITECTURE", "solution_architect"],
    ["TEST_PLAN", "acceptance_test_engineer"],
  ] as const) {
    const node = definition.nodes.find((candidate) => candidate.id === id);
    if (node?.kind !== "agent" || node.agent !== agent) {
      throw new Error(`FDE ${id} 必须由 ${agent} 执行`);
    }
  }
  const review = definition.nodes.find((node) => node.id === "REVIEW");
  if (review?.kind !== "agent" || review.agent !== "delivery_reviewer" || review.gate === null) {
    throw new Error("FDE REVIEW 必须由 delivery_reviewer 执行并保留质量门");
  }
  const humanAcceptance = definition.nodes.find((node) => node.id === "HUMAN_ACCEPTANCE");
  if (
    humanAcceptance?.kind !== "handler" ||
    humanAcceptance.mode !== NodeMode.HITL ||
    humanAcceptance.handler !== "engagement.human_acceptance" ||
    humanAcceptance.gate === null ||
    !humanAcceptance.gate.require.includes("signed == true") ||
    !humanAcceptance.gate.require.includes("decision_recorded == true") ||
    !humanAcceptance.gate.require.includes("package_bound == true")
  ) {
    throw new Error("FDE HUMAN_ACCEPTANCE 必须保留与精确包版本绑定的正式人工签字门");
  }
  const exported = definition.nodes.find((node) => node.id === "EXPORT");
  if (
    exported?.kind !== "handler" ||
    exported.handler !== "engagement.export" ||
    exported.gate === null ||
    !exported.gate.require.includes("human_decided == true")
  ) {
    throw new Error("FDE EXPORT 必须保留确定性交付门");
  }
}

assertFdeSafetyContract(FDE_DEFINITION);

export const FDE_ENGAGEMENT_AGENTS: readonly string[] = [...FDE_DEFINITION.requiredAgents];
export const FDE_CHECKPOINT_VERSION = FDE_DEFINITION.checkpointVersion;

/**
 * Build a fresh frozen front-line workflow.  The optional AgentLibrary retains the
 * historical fail-fast compatibility check; executable node specs still come from the
 * default built-in catalog, exactly as before this file-structure migration.
 */
export function buildFdeEngagementDag(
  agents?: AgentLibrary | null,
  opts: { checkpointSalt?: string } = {},
) {
  const compatibilityLibrary = agents ?? defaultAgents();
  for (const name of FDE_ENGAGEMENT_AGENTS) compatibilityLibrary.get(name);

  const checkpointVersion = opts.checkpointSalt
    ? `${FDE_CHECKPOINT_VERSION}+fork.${opts.checkpointSalt}`
    : FDE_CHECKPOINT_VERSION;
  return buildWorkflowDag(FDE_DEFINITION, defaultAgents(), { checkpointVersion });
}
